File: Examples/ChineseExample/JyutpingTransliterator.swift
//
//  JyutpingTransliterator.swift
//  ChineseExample
//
//  Cantonese romanisation. Faithfully extracted from the BiangBiang Hanzi
//  reference `hanziToJyutping(hanzi:)`: per-character dictionary lookup,
//  unknown characters preserved verbatim, tokens joined by single spaces.
//

import BiangBiangUI
import Foundation

public struct JyutpingTransliterator: Transliterator {
    public init() {}

    public func transliterate(_ scriptSpan: String) -> String {
        let dict = JyutpingDictionary.shared
        var tokens: [String] = []
        for scalar in scriptSpan {
            let key = String(scalar)
            if let reading = dict.reading(for: key) {
                tokens.append(reading)
            } else {
                tokens.append(key)
            }
        }
        return tokens.joined(separator: " ")
    }
}

File: android/examples/src/main/java/dev/veeso/biangbiangui/examples/chinese/JyutpingTransliterator.kt
package dev.veeso.biangbiangui.examples.chinese

import dev.veeso.biangbiangui.protocols.Transliterator

/**
 * Cantonese romanisation. Faithfully extracted from the BiangBiang Hanzi
 * reference Android `TextProcessor.hanziToJyutping(text)`: per-character
 * dictionary lookup, unknown characters preserved verbatim, tokens joined by
 * single spaces. The library's
 * [dev.veeso.biangbiangui.services.TextProcessingEngine] owns span detection
 * and spacing.
 *
 * Mirrors iOS `ChineseExample/JyutpingTransliterator`.
 */
class JyutpingTransliterator(private val dictionary: JyutpingDictionary) : Transliterator {

    override fun transliterate(scriptSpan: String): String {
        val out = StringBuilder()
        for (ch in scriptSpan) {
            val key = ch.toString()
            val reading = dictionary.reading(key) ?: key
            if (out.isNotEmpty()) out.append(' ')
            out.append(reading)
        }
        return out.toString()
    }
}
